text/0000-prefetch-hook.md

```diff
@@ -0,0 +1,79 @@
+- Start Date: 2015-10-01
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+The waterfall way in which Ember resolves a transition's model hooks can result in unnecessary blocking.
+Performance wins can be achieved by allowing routes to decouple initiating network requests from waiting for their completion.
+
+# Motivation
+
+Currently, the `model` hook is Ember's way for routes to request data.
+However, a route's `model` hook is not entered until it's parent route has resolved.
+This is suboptimal for nested application structures.
+Child routes are forced to wait for their parents to receive data before they may make requests for their own data.
+This occurs even when the child doesn't make use of its parents' data.
+
+This RFC proposes adding a new `Route#prefetch` hook, which is used to allow routes to make requests in parallel.
+Child routes that make use of `prefetch` will resolve faster since their data can resolve at the same time as (or before) their parents'.
+The semantics and ordering of `Route`'s existing model hooks (`beforeModel`, `model`, `afterModel`) are preserved.
+
+See this demo for a visualization: http://nickiaconis.github.io/ember-parallel-model-demo/
+
+# Detailed design
+
+A `prefetch` hook is added to `Route`.
+It takes the same parameters as the `model` hook.
+Like the `model` hook, it is not called if an object is passed to the transition.
+
+```javascript
+App.PostRoute = Ember.Route.extend({
+  prefetch(params) {
+    return Ember.$.get(`/api/posts/${params.id}`);
+  }
+});
+
+App.PostCommentsRoute = Ember.Route.extend({
+  prefetch(params, transition) {
+    return Ember.$.get(`/api/posts/${transition.params.post.id}/comments`);
+  }
+});
+```
+
+The default functionality of the `model` hook is modified to return the prefetched data if it exists.
```

Comment:
I'm unsure if we should modify the existing/default `model`. It is unclear to me that the relationship between prefetch and model is actually 1:1

given the following scenario:

```js
// routes/user.js
export default Route.extend({
  prefetch(params) {
    return RSVP.hash([
       user: this.store.find('user', params.user_id),
       friends: this.store.find('friends', { user_id: params.user_id }),
       somethingElse: ajax(...)
    })
  },

  // the default model functionality of `this.store.find('user', params.user_id)` will "just work" and the ED identify map would benefit from the existing fetch.
})
```

I would imagine, adding / removing prefetch request would be common, as the the UI changes or the data topology / size changes, tuning will occur.

In addition, query-params may or may not go through the pre-fetch if model refreshes.

If the model wants access to the prefetched data, it should likely be explicit.

```js
export default Route.extend({
  prefetch()  { ... }
  model() {
    return this.prefetched();
  }
})
``

```js
export default Route.extend({
  prefetch()  { ... }
  model() {
    return this.prefetched('other-route');
  }
})
````

It almost seems like `prefetched` concept could be symmetric to modelFor, but always returning a promise.

This would also allow an advanced user to create additional async dependencies while `prefetching`

maybe for example:

```js
export default Route.extend({
  prefetch() {
    // although this now creates a sequence, prefetch firing concurrently at the start of routing,
    // enables these to speed ahead, but truthfully support inter-route prefetching dependencies.
    return this.prefetched('otherRoute').then(other => {
      return this.store.find('user', other.id);
    })
  }
})
```